In [2]:
import time

import numpy as np
from numpy.ma.core import zeros, append
from pkg_resources import normalize_path
from typing_extensions import reveal_type

np.random.seed(42)

#Sentetik veri üretme

veri = np.column_stack([
    np.random.normal(0,1,1000),
    np.random.uniform(0,10,1000),
    np.random.exponential(2,1000),
    np.random.randint(1,100,1000),
    np.random.normal(5,2,1000)
])

sutun_isimleri="Normal_Dagilim,Duzgun_Dagilim,Ustel_Dagilim,Tam_Sayilar,Normal_Dagilim2"

np.savetxt('sentetik_veri.csv',veri,delimiter=',',header=sutun_isimleri,comments='',fmt='%.6f')

print("CSV dosyası başarı ile oluşturuldu...")
print ("----İçindeki Veriler----\n")

veriler = np.loadtxt('sentetik_veri.csv',delimiter=',',skiprows=1,dtype='float')
print(veriler)


CSV dosyası başarı ile oluşturuldu...
----İçindeki Veriler----

[[ 4.967140e-01  1.674826e+00  4.945360e-01  5.700000e+01  7.166546e+00]
 [-1.382640e-01  1.045678e+00  7.482500e-02  4.900000e+01  3.867187e+00]
 [ 6.476890e-01  6.364302e+00  2.286360e-01  4.500000e+01  3.841600e+00]
 ...
 [ 6.408430e-01  5.165726e+00  1.560729e+00  7.400000e+01  5.403088e+00]
 [-5.711790e-01  5.505800e-02  6.888156e+00  7.800000e+01  3.471908e+00]
 [ 5.725830e-01  7.664000e-02  1.377670e-01  5.800000e+01  2.870002e+00]]


In [3]:
import  numpy as np
data=np.loadtxt('sentetik_veri.csv',delimiter=',',skiprows=1,dtype='float')
tam_sayilar=data[:,3]

# Veri listesindeki veriler
print(f"Normalize edilmemiş ham veriler: {tam_sayilar}\n")

# Veri listesindeki normalizasyon işlemi
enb=tam_sayilar.max()
enk=tam_sayilar.min()
normalize_veri=((tam_sayilar-enk)/(enb-enk))
print (f"Normalizasyon işlemi yapılmış veri listesi: {normalize_veri}")

Normalize edilmemiş ham veriler: [57. 49. 45. 82. 66. 77. 45. 79. 20. 64. 93. 93. 67. 83.  3. 54. 78. 64.
 40. 59. 55. 62.  4. 60. 75. 95. 96. 91. 30.  2. 62. 49.  9. 92. 31. 98.
 61. 79. 48. 25. 36.  8. 73. 58. 45. 62. 39. 96. 58. 81. 99. 83. 64. 44.
 29. 59. 71. 92.  9. 92.  6. 29. 58. 17. 16. 87. 95.  3. 21. 64. 90. 36.
 86. 62. 42. 76. 42. 61. 13. 11. 54. 16.  8. 59. 57. 40. 59.  2. 94. 94.
 76. 10. 44. 11. 79. 80. 81. 90.  2. 50. 12. 90. 23. 18. 62. 96. 62. 38.
 67. 38. 67. 94. 61.  1. 94. 45. 96. 28. 21. 13. 27.  7. 22. 74. 53. 79.
 79.  4. 96. 23. 70. 88. 84. 84. 76. 33. 28.  8.  1. 37. 63. 22. 70. 21.
 96. 35. 79.  5. 60. 46. 50. 48.  6. 89. 56. 59. 29. 71. 77. 18.  2.  8.
 75. 97. 54. 38. 48. 27. 20. 12. 43. 61. 31. 91. 96. 61. 74. 67. 91. 35.
 95. 15. 16. 13. 35. 75. 59. 21. 54. 75. 30. 48. 35. 86. 38. 31. 78. 24.
 49. 96. 32. 55. 79. 58. 99.  3. 80. 69. 85. 10. 83. 25. 68. 61. 29. 19.
 43. 66. 74. 61. 43. 25. 20. 48. 83. 87. 24. 11. 98. 55. 28. 69. 63. 59.
  5. 51. 21. 12. 5

In [4]:
import numpy as np

data=np.loadtxt('sentetik_veri.csv',delimiter=',',skiprows=1,dtype='float')
ustel_dagilim=data[:,2]
medyan=np.median(ustel_dagilim)
#Eşikleme öncesi verilerin toplam standart sapması,medyan ve ortalama
print(f"Toplam standart sapma: {ustel_dagilim.std()}\n")
print(f"Veri grubu ortalamsı : {ustel_dagilim.mean()}\n")
print(f"Veri grubu medyanı: {medyan}\n")

#Eşikleme ile grafiğin ne yönlü olduğunu bulma
sola_mi_yatik =ustel_dagilim.mean() < medyan
print (f"Grafik sola doğru mu yatık : {sola_mi_yatik}\n")


Toplam standart sapma: 1.889704160207406

Veri grubu ortalamsı : 1.946503911

Veri grubu medyanı: 1.35341

Grafik sola doğru mu yatık : False



In [5]:
import numpy as np

data=np.loadtxt('sentetik_veri.csv',delimiter=',',skiprows=1,dtype='float')
duzgun_dagilim=data[:,1]

#Gruplandırılmamış veri grubu
print(duzgun_dagilim)

#Kategori kodlama işlemi
q1=np.percentile(duzgun_dagilim,25)
q2=np.percentile(duzgun_dagilim,50)
q3=np.percentile(duzgun_dagilim,75)

def kategorile(deger):
    if deger<q1:
        return "1.ceyrek"
    elif deger<q2:
        return "2.ceyrek"
    elif deger<q3:
        return "3.ceyrek"
    else:
         return "4.ceyrek"

kategorilenmis_veriler=np.vectorize(kategorile)(duzgun_dagilim)
print(f"Kategorilerine ayrılmış veriler :{kategorilenmis_veriler}")

[1.674826 1.045678 6.364302 7.064757 0.315861 9.362122 0.519713 5.412963
 7.090605 8.709691 7.140869 8.017281 3.394502 8.148251 0.801148 8.948167
 5.475924 8.172978 4.523183 6.435777 5.264027 7.315895 0.8163   0.603521
 2.471032 1.595447 8.717836 2.19214  9.758653 3.368958 1.821179 7.896985
 6.587078 4.981957 5.553636 7.192018 2.284547 9.963339 9.747932 6.503257
 1.995425 6.802282 0.721984 0.306525 2.576829 4.62623  8.682725 7.271691
 7.427065 4.254933 3.45935  3.710388 9.876496 0.401092 8.670315 5.786754
 4.386154 7.252577 4.866689 8.734232 9.007019 4.217209 2.768278 5.923503
 9.123633 2.106622 6.229666 6.315602 7.33113  1.315677 7.15825  9.090325
 1.796831 2.375433 9.713951 1.80977  8.543851 4.922779 2.472311 8.707499
 4.453053 5.148174 3.592334 5.929509 1.635239 3.910815 9.694123 2.581334
 6.567367 3.251901 7.734731 1.308737 9.69821  4.537895 2.360505 0.734967
 1.697579 5.197739 3.370032 8.288834 4.308875 2.487143 6.17145  7.067772
 1.670419 1.676192 0.366714 7.36402  6.638045 4.746

In [6]:
import numpy as np
a=np.array([
    [
        [0,2,4],
        [6,8,10],
        [12,14,16],
        [18,20,22]
    ],
    [
        [1,3,5],
        [7,9,11],
        [13,15,17],
        [19,22,23]
    ],
    [
        [1,1,1],
        [2,2,2],
        [3,3,3],
        [4,4,4]
    ]
])

print("Shape:", a.shape,"Dtype:",a.dtype,"ndim:",str(a.ndim))

z=np.zeros((2,3))
o=np.ones((3,3))
e=np.eye(3)
s=np.linspace(0,1,15)
A=np.arange(0,4,1)

print(A)
print(o)
print(e)
print(s)
print(z)
print(z.dtype)
print(z.size)
print(o.shape)
print("Shape:  ",A.shape,"Dtype: ",A.dtype,"ndim: ",str(A.ndim),"nbytes: ",A.nbytes)




Shape: (3, 4, 3) Dtype: int64 ndim: 3
[0 1 2 3]
[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
[0.         0.07142857 0.14285714 0.21428571 0.28571429 0.35714286
 0.42857143 0.5        0.57142857 0.64285714 0.71428571 0.78571429
 0.85714286 0.92857143 1.        ]
[[0. 0. 0.]
 [0. 0. 0.]]
float64
6
(3, 3)
Shape:   (4,) Dtype:  int64 ndim:  1 nbytes:  32


In [7]:
import  numpy as np
# 3d/2d dizilerin oluşturulması
a=np.array([
    [
        [0,2,4],
        [6,8,10],
        [12,14,16],
        [18,20,22]
    ],
    [
        [1,3,5],
        [7,9,11],
        [13,15,17],
        [19,22,23]
    ],
    [
        [1,1,1],
        [2,2,2],
        [3,3,3],
        [4,4,4]
    ]
])

b= np.array([
    [1,2,3,4,5],
    [6,7,8,9,10],
    [12,13,14,15,16],
])

In [8]:
# Dilimleme ornekleri
a[0, : ,:] #ilk derinliği alma

array([[ 0,  2,  4],
       [ 6,  8, 10],
       [12, 14, 16],
       [18, 20, 22]])

In [9]:
a[0,0,0]  # ilk elemanı alma

np.int64(0)

In [10]:
a[:,:,2].T   # Tensorun son sütunlarının transpoz matrisini alma

array([[ 4,  5,  1],
       [10, 11,  2],
       [16, 17,  3],
       [22, 23,  4]])

In [11]:
a[:,2,:]     # Tensorun 2. indexdeki satırlarını alma

array([[12, 14, 16],
       [13, 15, 17],
       [ 3,  3,  3]])

In [12]:
a[:,:,0]    # Tensonun ilk sutunlarını alma

array([[ 0,  6, 12, 18],
       [ 1,  7, 13, 19],
       [ 1,  2,  3,  4]])

In [13]:
b[:,0] # 2D dizimizin ilk sütununu  alma

array([ 1,  6, 12])

In [14]:
b[0,:]  #2D dizimizin ilk satırını alma

array([1, 2, 3, 4, 5])

In [15]:
b[1,1]  #2D dizimizin belirli konumdaki elemanı getirme

np.int64(7)

In [16]:
# Boolean Maske ile filtreleme

In [17]:
# Fermat sayıları is_fermat fonksiyonu ile boolean filtereleme
def is_fermat(x) :
    for i  in range(10):
        if x ==2**(2**i)+1:
         return True
    return False

# Dizi içindeki elemanların fermat olup olmadığını görme
def fermat_list() :
    for i in range(len(b)):
        for j in  range(len(b[i])):
            if is_fermat(b[i,j]):
               print("Fermat number",b[i,j])
            else:
                 print("Not Fermat number",b[i,j])


fermat_list()


Not Fermat number 1
Not Fermat number 2
Fermat number 3
Not Fermat number 4
Fermat number 5
Not Fermat number 6
Not Fermat number 7
Not Fermat number 8
Not Fermat number 9
Not Fermat number 10
Not Fermat number 12
Not Fermat number 13
Not Fermat number 14
Not Fermat number 15
Not Fermat number 16


In [18]:
# Z-score normalizasyonu (standartlaştırma), bir veri setindeki değerleri aynı ölçeğe getirmek için kullanılan istatistiksel bir yöntemdir. Satır bazlı z-score normalizasyonunda, her satır kendi içinde normalleştirilir.
 # z = (x - μ) / σ

In [19]:
 data=np.loadtxt('sentetik_veri.csv',delimiter=',',skiprows=1,dtype='float')


In [20]:
ort=np.mean(data,axis=1) #Satırlar boyunca ortalama al
print(ort)

[13.3665244 10.7698852 11.2164454 19.1930842 15.1198168 18.360817
 10.6596038 18.2844838  7.0361224 15.465552  20.908633  21.4839394
 15.3491894 19.2851332  1.2588536 13.9816382 17.3764468 16.5470362
 11.6493298 13.8675364 13.4975304 14.7324864  2.821627  13.2119112
 17.353865  20.3999224 21.623096  19.7380666  9.2460384  2.8294504
 14.1841024 12.663908   4.5761122 20.004277   8.1913468 21.4296144
 14.554201  18.364128  12.4778578  7.741541   8.9419646  4.886075
 15.9735544 13.5554112 10.5156428 13.957161  10.688863  21.9364366
 15.9654406 18.2767974 21.5929718 17.7500512 15.9688462  9.9709776
  9.2701156 15.6059434 16.7203242 21.3847462  3.8369654 20.8996942
  4.541818   7.487119  13.3598134  5.0863966  6.3040988 19.4566846
 21.4287148  2.878591   7.2700262 14.784509  20.6547888  9.4716024
 19.2166832 14.429427  12.1476826 17.5201748 10.9240022 13.5236746
  5.175788   5.1151918 13.6833778  5.927758   3.8279728 14.369709
 12.993372  10.3043802 16.0027022  1.8699962 20.4934046 20.966084

In [21]:
std=np.std(data,axis=1) # Satırlar boyunca standart sapma al
print(std)

[21.95602806 19.16831321 17.03931201 31.48252175 25.54597276 29.4951418
 17.23001242 30.43199459  6.95655774 24.44032304 36.15610659 35.87162464
 25.91170498 32.06379128  1.70585091 20.23632559 30.3858941  23.92059569
 14.57984029 22.75202903 20.83517711 23.77429356  2.97108953 23.51441416
 28.89328006 37.31172366 37.32613565 35.65877694 10.89184988  2.1031108
 23.96547586 18.36180433  3.21546485 36.069465   11.54376532 38.39379289
 23.34828563 30.55984142 18.19424069  9.00091601 13.60037808  3.59548564
 28.60024568 22.4591229  17.33404935 24.10750423 14.57418816 37.10733295
 21.20001631 31.46167675 38.71800237 32.65189405 24.27215256 17.10370005
 10.23052902 21.80191921 27.26878506 35.38939263  2.95399687 35.67556089
  3.10865022 10.92480492 22.42440407  6.44674352  5.86711927 33.81251759
 36.8681511   1.88451121  7.33220447 24.85638258 34.78073708 13.68166843
 33.49480178 23.79409537 15.63780583 29.2797228  15.80226116 23.80292253
  4.72126709  4.94712553 20.44299326  5.32618078  2.4

In [22]:
((data[:,0]-ort)/std) # normal_dağılımın z-score normalizasyonu

array([-0.58616296, -0.56907194, -0.62025723, -0.56126553, -0.60103289,
       -0.63044125, -0.52701011, -0.5756129 , -1.0789239 , -0.61058898,
       -0.59110488, -0.61189505, -0.58302715, -0.66113246, -1.74913973,
       -0.7187039 , -0.60519127, -0.67861141, -0.86128199, -0.67158144,
       -0.5774792 , -0.629178  , -0.92696601, -0.62245477, -0.61946058,
       -0.54377009, -0.6101379 , -0.54299026, -0.90404087, -1.48406085,
       -0.6169629 , -0.58881087, -1.42735481, -0.58392848, -0.63833607,
       -0.58995105, -0.61440644, -0.6650492 , -0.75881396, -0.83821247,
       -0.60318158, -1.31128517, -0.56255469, -0.616966  , -0.69194246,
       -0.60881478, -0.76501702, -0.56267355, -0.73687786, -0.63696025,
       -0.54932813, -0.55540831, -0.68579695, -0.54720918, -0.805346  ,
       -0.67309044, -0.64394296, -0.61300736, -1.18676578, -0.55848174,
       -1.61516788, -0.70232632, -0.64510737, -0.9745391 , -0.93599134,
       -0.53531786, -0.58317882, -0.99498373, -0.94219825, -0.62

In [23]:
((data[:,1]-ort)/std) # düzgün_dağılımın z-score normalizasyonu

array([-5.32505167e-01, -5.07306360e-01, -2.84761697e-01, -3.85240017e-01,
       -5.79502528e-01, -3.05090752e-01, -5.88501654e-01, -4.22960144e-01,
        7.83183322e-03, -2.76422738e-01, -3.80786686e-01, -3.75412559e-01,
       -4.61362439e-01, -3.47335164e-01, -2.68315125e-01, -2.48734444e-01,
       -3.91646293e-01, -3.50077327e-01, -4.88767137e-01, -3.26641610e-01,
       -3.95173190e-01, -3.11958435e-01, -6.74946676e-01, -5.36198355e-01,
       -5.15096693e-01, -5.03983026e-01, -3.45743265e-01, -4.92050712e-01,
        4.70640530e-02,  2.56528377e-01, -5.15863882e-01, -2.59610816e-01,
        6.25404380e-01, -4.16483028e-01, -2.28496572e-01, -3.70830682e-01,
       -5.25505564e-01, -2.74896355e-01, -1.50043404e-01, -1.37573109e-01,
       -5.10760772e-01,  5.32948034e-01, -5.33267111e-01, -5.89911114e-01,
       -4.57989570e-01, -3.87055039e-01, -1.37650069e-01, -3.95198049e-01,
       -4.02753256e-01, -4.45680773e-01, -4.68351172e-01, -4.29980055e-01,
       -2.51001644e-01, -

In [24]:
data_nan=np.loadtxt('sentetik_eksik_veri.csv',delimiter=',',skiprows=1,dtype='float')

In [25]:
((data[:,2]-ort)/std) # üstel_dağılımın z-score normalizasyonu

array([-0.58626216, -0.55795521, -0.64485053, -0.58335553, -0.46484493,
       -0.56495155, -0.53522804, -0.57799862, -0.23960951, -0.59139259,
       -0.57324731, -0.54539351, -0.58303679, -0.57504776,  0.09221814,
       -0.48601492, -0.52708012, -0.59045249, -0.12190112, -0.59711538,
       -0.61699036, -0.57746546, -0.58088421, -0.51683963, -0.43575426,
       -0.47963926, -0.52913576, -0.53276523, -0.5301098 ,  1.60496851,
       -0.44819249, -0.68951693, -0.7938831 , -0.55169377, -0.65798226,
       -0.55495623, -0.54404645, -0.52499101, -0.67244795, -0.75248419,
       -0.56788735, -1.10277398, -0.54949026, -0.55521234, -0.44615125,
       -0.57284449, -0.72194128, -0.56754646, -0.48491716, -0.5107199 ,
       -0.49905464, -0.51609466, -0.58370234, -0.58178123, -0.70850536,
       -0.49149991, -0.54362034, -0.4873375 , -0.4273706 , -0.57862463,
        0.03635983, -0.67274528, -0.52767571, -0.74142466, -1.07046534,
       -0.54300019, -0.56225995, -0.8211811 , -0.75670478, -0.59

In [26]:
((data[:,3]-ort)/std) # tam-sayıların z-score normalizasyonu

array([ 1.98731189,  1.99444335,  1.98268302,  1.99497729,  1.99171054,
        1.98809632,  1.99305696,  1.99512116,  1.86354776,  1.98583496,
        1.99389187,  1.99366662,  1.99333894,  1.98712829,  1.02069084,
        1.9775508 ,  1.99512158,  1.98377015,  1.94451171,  1.98366763,
        1.99194225,  1.98817742,  0.39661309,  1.98976204,  1.99513987,
        1.99937366,  1.99262267,  1.9984402 ,  1.90545792, -0.39439215,
        1.99519917,  1.9788955 ,  1.37581594,  1.99602969,  1.97584173,
        1.99434283,  1.9892595 ,  1.98416841,  1.95238388,  1.9174114 ,
        1.98950612,  0.8660652 ,  1.99391454,  1.97891026,  1.98939997,
        1.99285826,  1.94255328,  1.99592796,  1.98276071,  1.99363826,
        1.9992516 ,  1.99835111,  1.97885843,  1.98957081,  1.92853022,
        1.99037783,  1.99054251,  1.99537908,  1.74781316,  1.99296953,
        0.46907239,  1.96917759,  1.99069667,  1.84800332,  1.65258294,
        1.99758315,  1.99552413,  0.06442466,  1.87255741,  1.97

In [27]:
((data[:,4]-ort)/std) # normal_dağılım2'nin  z-score normalizasyonu

array([-0.2823816 , -0.36010984, -0.43281357, -0.46511621, -0.3463302 ,
       -0.48761278, -0.34231715, -0.41854949, -0.55284618, -0.50743065,
       -0.44875299, -0.4609655 , -0.36591256, -0.40361291,  0.90454587,
       -0.52409753, -0.4712039 , -0.36462893, -0.47256147, -0.38832921,
       -0.4022995 , -0.46957553,  1.78618381, -0.31426929, -0.42482833,
       -0.47198129, -0.50760575, -0.430634  , -0.5183713 ,  0.01695612,
       -0.4141799 , -0.44095688,  0.22001758, -0.44392441, -0.45102682,
       -0.47860487, -0.30530104, -0.51923185, -0.37107857, -0.18914164,
       -0.30767642,  1.01504591, -0.34860247, -0.21682081, -0.39331668,
       -0.42414395, -0.31794491, -0.4705099 , -0.35821244, -0.40027734,
       -0.48251766, -0.49686809, -0.4583575 , -0.30105904, -0.35605037,
       -0.37540559, -0.35066092, -0.49570074, -0.48226334, -0.51486047,
       -0.3266437 , -0.29479538, -0.34559177, -0.26188906, -0.12669161,
       -0.40613967, -0.4378312 , -0.07208076, -0.181988  , -0.22

In [28]:
# Eksik veri (NaN) enjekte etme


In [29]:
def inject_random_nan(data_nan,missing_rate=0.1): # Eksik veri eklenecek grup,kayıp oranı 0.1 => %10 kayıp oranı
    data_copy=data_nan.copy().astype(float) # veri kopyalama
    mask=np.random.random(data_nan.shape)<missing_rate # mask oluşturma
    data_copy[mask]=np.nan # belirtilen yerlere nan enjekte etme
    return data_copy,mask



In [30]:
data_whit_nan,nan_mask =inject_random_nan(data_nan,missing_rate=0.15) #fonksiyonu %15 kayıpla dene
print(nan_mask)
print(data_whit_nan)

[[False False False False False]
 [ True False False False False]
 [False False False False False]
 ...
 [False False False  True  True]
 [ True False False  True False]
 [False  True False False False]]
[[4.967140e-01 1.674826e+00 4.945360e-01 5.700000e+01 7.166546e+00]
 [         nan 1.045678e+00 7.482500e-02 4.900000e+01 3.867187e+00]
 [6.476890e-01 6.364302e+00 2.286360e-01 4.500000e+01 3.841600e+00]
 ...
 [6.408430e-01 5.165726e+00 1.560729e+00          nan          nan]
 [         nan 5.505800e-02 6.888156e+00          nan 3.471908e+00]
 [5.725830e-01          nan 1.377670e-01 5.800000e+01 2.870002e+00]]


In [31]:
# Her sutunun kendi ortalaması ile NaN verilerini doldurma
def NaN_doldur(data_nan):
    data_filled=data_nan.copy() # Data kopyalama
    col_means=np.nanmean(data_filled,axis=0) # kolon bazında axis=0 değeri kullanılır

    for idx in range(data_nan.shape[1]):  # .shape[idx]: n boyutlu bir dizinin belirtilen indeksdeki boyutunu getirir
        nan_mask=np.isnan(data_nan[:,idx]) # NaN olan yerleri  belirledik
        data_filled[nan_mask,idx]=col_means[idx] # NaN yerine o sütünün ortalaması ile doldurduk.

    return data_filled




In [32]:
doldurulmus_data=NaN_doldur(data_nan)
print("Sütun ortalaması ile doldurulmuş data :")
print(doldurulmus_data)

Sütun ortalaması ile doldurulmuş data :
[[ 4.967140e-01  1.674826e+00  4.945360e-01  5.700000e+01  7.166546e+00]
 [-1.382640e-01  1.045678e+00  7.482500e-02  4.900000e+01  3.867187e+00]
 [ 6.476890e-01  6.364302e+00  2.286360e-01  4.500000e+01  3.841600e+00]
 ...
 [ 6.408430e-01  5.165726e+00  1.560729e+00  7.400000e+01  5.403088e+00]
 [-5.711790e-01  5.505800e-02  6.888156e+00  7.800000e+01  3.471908e+00]
 [ 5.725830e-01  7.664000e-02  1.377670e-01  5.800000e+01  2.870002e+00]]


In [33]:
# Aykırı değer bulma
doldurulmus_data.max()

np.float64(99.0)

In [34]:
doldurulmus_data.min()

np.float64(-3.241267)

In [35]:
doldurulmus_data.mean() # Aykırı değer sınırlandırması için ortalama değer ve medyan bulunması

np.float64(12.644455065400003)

In [36]:
np.median(doldurulmus_data)

np.float64(3.7385295)

In [37]:
clipped=np.clip(doldurulmus_data,0,15) # Aykırı değer sınırlandırılması
print(clipped)

[[ 0.496714  1.674826  0.494536 15.        7.166546]
 [ 0.        1.045678  0.074825 15.        3.867187]
 [ 0.647689  6.364302  0.228636 15.        3.8416  ]
 ...
 [ 0.640843  5.165726  1.560729 15.        5.403088]
 [ 0.        0.055058  6.888156 15.        3.471908]
 [ 0.572583  0.07664   0.137767 15.        2.870002]]


In [38]:
clipped.mean()  #Sonraki ortalama ve medyan değerleri

np.float64(5.2835600074)

In [39]:
np.median(clipped)

np.float64(3.7385295)

# Gerekçe
Tam sayılar sütunu verinin tamamının dağılımının aksine büyük sayılardan oluşması nedeniyle ortlamanın gerçek veri dağılımına yakın olamaması ile sonuçlanıyordu

In [40]:
# Random kare matris üretme
np.random.seed(42)
A=np.random.rand(3,3)

In [41]:
# Vektör oluşturma
b=np.array([1.0,2.0,3.0]).reshape(3,1)
b

array([[1.],
       [2.],
       [3.]])

In [42]:
# x'i bulma
x= np.linalg.solve(A,b)
x

array([[  7.77085163],
       [ 48.18800129],
       [-65.19660231]])

In [43]:
# cond(A) = ||A|| × ||A⁻¹||

In [44]:
norm_type=2  # İkinci dereceden norm alınacak
norm_A=np.linalg.norm(A,ord=norm_type) # Norm alma
norm_A

np.float64(1.666728164305919)

In [45]:
inv_A=np.linalg.inv(A) # A 'nın Tersini alma
inv_A

array([[ -1.72626243,   2.61222492,   1.42422141],
       [-14.65094529,   7.62717264,  15.86153376],
       [ 21.2780693 , -11.24277772, -21.32970539]])

In [46]:
norm_inv_A=np.linalg.norm(inv_A,ord=norm_type) # Norm alma
norm_inv_A

np.float64(39.58414547195899)

In [47]:
kosul_sayisi= norm_A*norm_inv_A # Formul ile manuel koşul sayısı bulma
int(kosul_sayisi)

65

In [48]:
numpy_kosul_sayisi=np.linalg.cond(A,p=norm_type) # Numpy kütüphanesi ile koşul sayısı bulma
int(numpy_kosul_sayisi)

65

Yukarıdaki lineer sistemin çöümünün sayısal(nümerik) analiz standartlarına göre analiz ettiğimizde koşul sayısının 65 sayısı olması 'Endüstride kabul görmüş koşul sayısı eşik değerleri' kriterleri bakımından çok iyi bir sonuç olduğu söylenebilir.yüksek hassasiyetle güvenilir çözüm sağlaması bakımından oldukça tutarlı bir çözümdür.

In [49]:
# time.perf_counter()  ile problemlerin hız ve performans testleri

In [50]:
import math

In [51]:
# Klasik standart sapma formülünün python döngüsü ile yapılması
def standart_sapma(veriler):
    ortalama = sum(veriler) / len(veriler)

    karelerin_farki = []
    for veri in veriler:
        karelerin_farki.append((veri - ortalama) ** 2)

    varyans = sum(karelerin_farki) / len(veriler)
    return math.sqrt(varyans)



In [52]:
# Saf python döngüsü ile performans ve hız testi 1.adım
import time
baslangic= time.perf_counter()
sonuc=standart_sapma(data[:,1])
bitis= time.perf_counter()
print(f"İkinci sütunun standart sapması: {sonuc:.2f}")
print(f"Hesaplama süresi: {bitis - baslangic:.6f} saniye")

İkinci sütunun standart sapması: 2.88
Hesaplama süresi: 0.006987 saniye


In [53]:
# Numpy kütüphanesi ile performans ve hız testi 1.adım
baslangic= time.perf_counter()
sonuc=np.std(data[:,1])
bitis= time.perf_counter()
print(f"İkinci sütunun standart sapması: {sonuc:.2f}")
print(f"Hesaplama süresi: {bitis - baslangic:.6f} saniye")

İkinci sütunun standart sapması: 2.88
Hesaplama süresi: 0.000190 saniye


In [54]:
# Saf python döngüsü ile performans ve hız testi 2.adım
import time
baslangic= time.perf_counter()
sonuc=standart_sapma(data[:,1])
bitis= time.perf_counter()
print(f"İkinci sütunun standart sapması: {sonuc:.2f}")
print(f"Hesaplama süresi: {bitis - baslangic:.6f} saniye")

İkinci sütunun standart sapması: 2.88
Hesaplama süresi: 0.000420 saniye


In [55]:
# Numpy kütüphanesi ile performans ve hız testi 2.adım
baslangic= time.perf_counter()
sonuc=np.std(data[:,1])
bitis= time.perf_counter()
print(f"İkinci sütunun standart sapması: {sonuc:.2f}")
print(f"Hesaplama süresi: {bitis - baslangic:.6f} saniye")

İkinci sütunun standart sapması: 2.88
Hesaplama süresi: 0.000188 saniye


In [56]:
# Saf python döngüsü ile performans ve hız testi 3.adım
import time
baslangic= time.perf_counter()
sonuc=standart_sapma(data[:,1])
bitis= time.perf_counter()
print(f"İkinci sütunun standart sapması: {sonuc:.2f}")
print(f"Hesaplama süresi: {bitis - baslangic:.6f} saniye")

İkinci sütunun standart sapması: 2.88
Hesaplama süresi: 0.000427 saniye


In [57]:
# Numpy kütüphanesi ile performans ve hız testi 3.adım
baslangic= time.perf_counter()
sonuc=np.std(data[:,1])
bitis= time.perf_counter()
print(f"İkinci sütunun standart sapması: {sonuc:.2f}")
print(f"Hesaplama süresi: {bitis - baslangic:.6f} saniye")

İkinci sütunun standart sapması: 2.88
Hesaplama süresi: 0.000188 saniye


# Hız Analizi: Saf Python Döngüsü vs. NumPy Metodu(Standart Sapma)

| Deneme | Saf Python Döngüsü (saniye) | NumPy Metodu (saniye) |
| :----: | :-------------------------: | :--------------------: |
|   i    |           0.000654          |        0.000242        |
|   ii   |           0.000667          |        0.000211        |
|  iii   |           0.000519          |        0.000266        |

---

Bu tabloya göre, **NumPy kütüphanesi** kullanılarak oluşturulan çözüm, saf Python döngüsüne kıyasla önemli ölçüde daha hızlıdır. Özellikle ileri düzey matematiksel hesaplamalar ve büyük veri setleriyle çalışırken NumPy, daha verimli bir seçenek sunar.

Sürekli olarak farklı çıkan süreler, işlemcinin her bir çalıştırmada farklı durum ve kaynakları kullanmasından kaynaklanabilir. Bu durum, arka planda çalışan diğer işlemler ve sistem kaynaklarının anlık kullanımı gibi faktörlere bağlı olabilir.

In [58]:
# Öklid Mesafesi Problemi

In [59]:
# Saf python döngüsü
def oklid_mesafe(v1,v2):
    kare_toplam_fark=0
    for i in range(len(v1)):
        kare_toplam_fark+=(v1[i]-v2[i])**2
    return math.sqrt(kare_toplam_fark)

In [60]:
# Numpy Kütüphanesi
def oklid_mesafe_numpy(v1,v2):
    return np.linalg.norm(np.array(v1)-np.array(v2))


In [61]:
v1=[1,2,3]
v2=[4,5,6]


In [62]:
baslangic=time.perf_counter()
sonuc=oklid_mesafe(v1,v2)
bitis= time.perf_counter()
print(f"İki vektör arası öklid mesafesi: {sonuc:.2f}")
print(f"Hesaplama süresi: {bitis - baslangic:.6f} saniye")

İki vektör arası öklid mesafesi: 5.20
Hesaplama süresi: 0.000040 saniye


In [63]:
baslangic=time.perf_counter()
sonuc=oklid_mesafe_numpy(v1,v2)
bitis= time.perf_counter()
print(f"İki vektör arası öklid mesafesi: {sonuc:.2f}")
print(f"Hesaplama süresi /numpy: {bitis - baslangic:.6f} saniye")

İki vektör arası öklid mesafesi: 5.20
Hesaplama süresi /numpy: 0.001751 saniye


In [64]:
baslangic=time.perf_counter()
sonuc=oklid_mesafe(v1,v2)
bitis= time.perf_counter()
print(f"İki vektör arası öklid mesafesi: {sonuc:.2f}")
print(f"Hesaplama süresi: {bitis - baslangic:.6f} saniye")

İki vektör arası öklid mesafesi: 5.20
Hesaplama süresi: 0.000033 saniye


In [65]:
baslangic=time.perf_counter()
sonuc=oklid_mesafe_numpy(v1,v2)
bitis= time.perf_counter()
print(f"İki vektör arası öklid mesafesi: {sonuc:.2f}")
print(f"Hesaplama süresi /numpy: {bitis - baslangic:.6f} saniye")

İki vektör arası öklid mesafesi: 5.20
Hesaplama süresi /numpy: 0.000088 saniye


In [66]:
baslangic=time.perf_counter()
sonuc=oklid_mesafe(v1,v2)
bitis= time.perf_counter()
print(f"İki vektör arası öklid mesafesi: {sonuc:.2f}")
print(f"Hesaplama süresi: {bitis - baslangic:.6f} saniye")

İki vektör arası öklid mesafesi: 5.20
Hesaplama süresi: 0.000035 saniye


In [67]:
baslangic=time.perf_counter()
sonuc=oklid_mesafe_numpy(v1,v2)
bitis= time.perf_counter()
print(f"İki vektör arası öklid mesafesi: {sonuc:.2f}")
print(f"Hesaplama süresi /numpy: {bitis - baslangic:.6f} saniye")

İki vektör arası öklid mesafesi: 5.20
Hesaplama süresi /numpy: 0.000088 saniye


# Hız Analizi: Saf Python Döngüsü vs. NumPy Metodu(Öklid Measfesi)

| Deneme | Saf Python Döngüsü (saniye) | NumPy Metodu (saniye) |
| :----: |:---------------------------:|:---------------------:|
|   i    |          0.000040           |       0.001751        |
|   ii   |          0.000033           |       0.000088        |
|  iii   |           0.000035          |       0.000088        |

---

Bu tabloya göre, **NumPy kütüphanesi** kullanılarak oluşturulan çözüm, saf Python döngüsüne kıyasla önemli ölçüde daha yavaştır. Bu aksine orneğin sebebi ise np.linalg modulu ile lineer cebir metodları ile çözülmeye çalışılmasıdır.Bu genellik, küçük boyutlar için daha yavaş olmasına neden olabilir,burada saf python döngüsü daha hızlı sonuç vermektedir.Bu vektörler np.array ile oluşturulmadığından np.array dönüşümü ek bir maliyet getirir.Genel olarak büyük boyutlu ve yüksek işlem gücü gerektiren problemlerde Numpy kütüphanesi daha başarılıdır.

Sürekli olarak farklı çıkan süreler, işlemcinin her bir çalıştırmada farklı durum ve kaynakları kullanmasından kaynaklanabilir. Bu durum, arka planda çalışan diğer işlemler ve sistem kaynaklarının anlık kullanımı gibi faktörlere bağlı olabilir.